crates/notion-core/src/inventory/mod.rs

```diff
@@ -108,32 +110,49 @@ impl Inventory {
         })
     }
 
-    /// Fetches a Node version matching the specified semantic versioning requirements.
-    pub fn fetch_node(&mut self, matching: &VersionSpec, config: &Config) -> Fallible<Fetched<NodeVersion>> {
-        let distro = self.node.resolve_remote(matching, config.node.as_ref())?;
-        let fetched = distro.fetch(&self.node).unknown()?;
+    /// Fetches a Tool version matching the specified semantic versioning requirements.
+    pub fn fetch(&mut self, _toolspec: &ToolSpec, _config: &Config) -> Fallible<Fetched<DistroVersion>> {
```

Comment:
If the idea is to have this be a general-purpose tool `fetch` command, the return type should probably support all of those tools, so this `DistroVersion` will probably need to support all the different types of tools and their associated versions.

Related: It would be nice if we could type this in a way that whatever tool you request, you are required to get that tool's `Version` type back. Right now it looks like you could call `fetch(ToolSpec::Node)` and get a return value that is `DistroVersion::Yarn`, which wouldn't make sense. I'm not sure exactly how to type that, but I feel like it should be possible.

Reply:
Good idea, it would be nice to use generics to do `fetch(ToolSpec<T>) -> DistroVersion<T>`, something like that. And I'm planning to extend DistroVersion to cover Node/Yarn/Npm/Npx/Package, so that should be enough to cover the things that can be fetched.

Reply:
I couldn't figure out how to make `fetch` generic for this, because it calls `inventory.fetch()` whcih is not generic-alizable as far as I can tell? Maybe @dherman has some thoughts on this.